### PR TITLE
Added a fibonacci algorithm which uses memoization

### PR DESCRIPTION
**Pull request by JDeepD**
### Describe your change:

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by JDeepD** [comment]
I dont know why the build is failing for my PR. It says :

```
Successfully installed types-setuptools-57.4.2 types-toml-0.10.1
strings/check_anagrams.py:32: error: Need type annotation for "count"
Installing missing stub packages:
/opt/hostedtoolcache/Python/3.9.7/x64/bin/python -m pip install types-setuptools types-toml


Found 1 error in 1 file (checked 677 source files)
Error: Process completed with exit code 1.
```

As per error, I think [this](https://github.com/TheAlgorithms/Python/blob/868c2fa0a8e1f51ba32e7622990a9259a8740604/strings/check_anagrams.py#L32) `count` variable should get a type hint.

File: maths/fibonacci_sequence_memoization.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+cache: dict[int, int] = {}
```
Comment (cclauss) [inline maths/fibonacci_sequence_memoization.py:1]: Prefill the cache...
```suggestion
cache: dict[int, int] = {1: 1, 2: 1}
```

File: maths/fibonacci_sequence_memoization.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    """
+    if num < 1:
+        raise ValueError("Enter positive integer")
+    elif num in cache:
+        return cache[num]
+    elif num in [1, 2]:
+        value = 1
+    else:
+        value = fibo(num - 1) + fibo(num - 2)
+
```
Comment (cclauss) [inline maths/fibonacci_sequence_memoization.py:23]:
```suggestion
    value = fibo(num - 1) + fibo(num - 2)
```

**Review by cclauss — CHANGES_REQUESTED** [review]

**Comment by pandrey2003** [comment]
@JDeepD btw, why is it chosen to define cache at the module level? Classical memoization technique enables creating an outer and inner function, something like:
```
def fibo() -> Callable:
    cache: dict[int, int] = {1: 1, 2: 1}

    def func_(n: int) -> int:
        nonlocal cache
        ...

    return func_
```

... and then we do something like:
```
fibonacci = fibo()
fibonacci(5)
```

The code in this PR is perfectly workable, but what about showcasing the "classical" ways of solving the problems?

**Comment by JDeepD** [comment]
> @JDeepD btw, why is it chosen to define cache at the module level? Classical memoization technique enables creating an outer and inner function, something like:
> 
> ```
> def fibo() -> Callable:
>     cache: dict[int, int] = {1: 1, 2: 1}
> 
>     def func_(n: int) -> int:
>         nonlocal cache
>         ...
> 
>     return func_
> ```
> 
> ... and then we do something like:
> 
> ```
> fibonacci = fibo()
> fibonacci(5)
> ```
> 
> The code in this PR is perfectly workable, but what about showcasing the "classical" ways of solving the problems?

Yeah sorry, I will change it right away.

**Comment by JDeepD** [comment]
> ```
> fibonacci = fibo()
> fibonacci(5)
> ```

Rather than creating another object, wouldn't it be better if we just pass the `num` from `fibonacci` to `fibo` directly?
Something like:

```
def fibo(num: int) -> int: 
       cache: dict[int, int] = {}
       def fibo(num:int) -> int:
           ...
       return fibo(num) 


```

**Comment by JDeepD** [comment]
Wait I messed something while pushing

**Comment by pandrey2003** [comment]
Hmm, quite frankly, I never saw such an implementation. My checks led me to one classical book about competitive programming ([link](https://cses.fi/book/book.pdf)), where we can see such an implementation of memoization in C++:

```
/* ready[x] indicates whether the value of solve(x) has been calculated,
and if it is, value[x] contains this value.
The constant N has been chosen so that all required values fit in the arrays. */
bool ready[N];
int value[N];

int solve(int x) {
    // the function per se
    if (x < 0) return INF;
    if (x == 0) return 0;
    if (ready[x]) return value[x];
    int best = INF;
    for (auto c : coins) {
       best = min(best, solve(x-c)+1);
    }
    value[x] = best;
    ready[x] = true;
    return best;
}
```

This piece of code is structurally implemented as your PR: cache is defined at the module level. However, this example shows how to write the code in the fastest way (because writing inner and outer functions takes time in programming contests).

However, if we look at some other [example](https://codeburst.io/understanding-memoization-in-3-minutes-2e58daf33a19), we can see we create an outer and an inner function in all cases: because we re-use this code and it should be written well.

**Comment by JDeepD** [comment]
Yeah, I also think putting the `cache`  outside the function is not good practice.Nevertheless, I have somehow managed to mess up the commit(It disappeared from remote smh). I will make a new PR shortly...

**Comment by pandrey2003** [comment]
OK, I'll take a look at it then :)

**Comment by JDeepD** [comment]
Thnx. The new PR is here #5856

**Pull request closed without merging**